### PR TITLE
ENH: Allow dict as the argument to Index.map (GH13517)

### LINKED ISSUE #13517: Index.map should allow dict as argument
Series.map works when passed a dict, but Index.map doesn't:

``` .python
>>> s = pd.Series(['a', 'b', 'c', 'd'], index=['a', 'b', 'c', 'd'])
>>> d = {'a': 0, 'b': 1, 'c': 2, 'd': 3}
>>> s.map(d)
a    0
b    1
c    2
d    3
dtype: int64
>>> s.index.map(d)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/evanpw/Workspace/pandas/pandas/indexes/base.py", line 2305, in map
    return self._arrmap(self.values, mapper)
  File "pandas/src/generated.pyx", line 2356, in pandas.algos.arrmap_object (pandas/algos.c:70221)
TypeError: 'dict' object is not callable
```

PR to follow shortly

### PR DESCRIPTION
**Pull request by evanpw**
- [X] closes #13517
- [X] tests added / passed
- [X] passes `git diff upstream/master | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/indexes/base.py [hunk 1]
```diff
@@ -2295,14 +2295,17 @@ def map(self, mapper):
...
-        mapper : callable
+        mapper : callable or dict
             Function to be applied.
 
         Returns
         -------
         applied : array
         """
-        return self._arrmap(self.values, mapper)
+        if isinstance(mapper, dict):
```
Comment (MaximilianR) [inline pandas/indexes/base.py]: `is_dict_like` covers a wider range of mappings

Comment (evanpw) [inline pandas/indexes/base.py]: Good point. Fixed.

**Comment by jreback** [comment]
I would ideally like to merge the ``Series.map`` and ``Index.map`` impl and move it to ``Base.map`` like most other methods. You can move tests to ``tests/test_base/py:TestIndexOps``. This will make these more consistent.

**Comment by jreback** [comment]
can you rebase / update?

**Comment by jreback** [comment]
closing, but pls reopen / comment if you can continue

**Pull request closed without merging**